src/core/mac/mac_frame.cpp

```diff
@@ -734,40 +764,78 @@ void Frame::SetPayloadLength(uint16_t aLength)
 
 uint8_t Frame::SkipSecurityHeaderIndex(void) const
 {
-    uint8_t  index = 0;
+    uint8_t  index = kFcfSize + kDsnSize; // Frame Control field and Sequence Number field
     uint16_t fcf;
 
-    // Frame Control
-    index += kFcfSize;
-    // Sequence Number
-    index += kDsnSize;
-
     VerifyOrExit((index + GetFcsSize()) <= GetPsduLength(), index = kInvalidIndex);
 
     fcf = GetFrameControlField();
+    index += SkipAddrField(fcf);
 
-    // Destination PAN + Address
+    // Security Control + Frame Counter + Key Identifier
+    if ((fcf & kFcfSecurityEnabled) != 0)
+    {
+        VerifyOrExit(index < GetPsduLength(), index = kInvalidIndex);
+        uint8_t securityControl = *(GetPsdu() + index);
+
+        index += kSecurityControlSize + kFrameCounterSize;
+
+        switch (securityControl & kKeyIdModeMask)
+        {
+        case kKeyIdMode0:
+            index += kKeySourceSizeMode0;
+            break;
+
+        case kKeyIdMode1:
+            index += kKeySourceSizeMode1 + kKeyIndexSize;
+            break;
+
+        case kKeyIdMode2:
+            index += kKeySourceSizeMode2 + kKeyIndexSize;
+            break;
+
+        case kKeyIdMode3:
+            index += kKeySourceSizeMode3 + kKeyIndexSize;
+            break;
+        }
+    }
+
+exit:
+    return index;
+}
+
+uint8_t Frame::SkipAddrField(const uint16_t &fcf) const
```

Comment:
nit: rename
```suggestion
uint8_t Frame::SkipAddrFields(const uint16_t &fcf) const
```

Reply:
Well, I'm not sure. Here is my thought: We had a function called `GetFrameControlField`. In a frame control field there are many fields, but they are as a whole called frame control field. So I think here the address field refers to the whole address thing.